engine/engine.go

```diff
@@ -92,84 +91,212 @@ func (e *compatibilityEngine) SetQueryLogger(l promql.QueryLogger) {
+		for _, s := range series {
+			if len(s.Points) == 0 {
+				continue
+			}
+			resultMatrix = append(resultMatrix, s)
+		}
+		sort.Sort(resultMatrix)
+		ret.Value = resultMatrix
+		return ret
 	}
 
-	return newInstantQuery(e.logger, plan, expr, ts), nil
+	var result parser.Value
+	switch q.expr.Type() {
+	case parser.ValueTypeMatrix:
+		result = promql.Matrix(series)
+	case parser.ValueTypeVector:
+		// Convert matrix with one value per series into vector.
+		vector := make(promql.Vector, 0, len(resultSeries))
+		for i := range series {
+			if len(series[i].Points) == 0 {
+				continue
+			}
+			// Point might have a different timestamp, force it to the evaluation
+			// timestamp as that is when we ran the evaluation.
+			vector = append(vector, promql.Sample{
+				Metric: series[i].Metric,
+				Point: promql.Point{
+					V: series[i].Points[0].V,
+					T: q.ts.UnixMilli(),
+				},
+			})
+		}
+		result = vector
+	case parser.ValueTypeScalar:
+		result = promql.Scalar{V: series[0].Points[0].V, T: q.ts.UnixMilli()}
+	default:
+		panic(errors.Newf("new.Engine.exec: unexpected expression type %q", q.expr.Type()))
+	}
+
+	ret.Value = result
+	return ret
 }
 
-func (e *engine) NewRangeQuery(q storage.Queryable, _ *promql.QueryOpts, qs string, start, end time.Time, interval time.Duration) (promql.Query, error) {
-	expr, err := parser.ParseExpr(qs)
-	if err != nil {
-		return nil, err
+func newErrResult(r *promql.Result, err error) *promql.Result {
+	if r == nil {
+		r = &promql.Result{}
+	}
+	if r.Err == nil && err != nil {
+		r.Err = err
+	}
+	return r
+}
+
+func (q *compatibilityQuery) Statement() parser.Statement { return nil }
+
+func (q *compatibilityQuery) Stats() *stats.Statistics { return &stats.Statistics{} }
+
+func (q *compatibilityQuery) Close() { q.Cancel() }
+
+func (q *compatibilityQuery) String() string { return q.expr.String() }
+
+func (q *compatibilityQuery) Cancel() {
+	if q.cancel != nil {
+		q.cancel()
+		q.cancel = nil
 	}
+}
+
+func (e *engine) SetQueryLogger(l promql.QueryLogger) {
+	e.logger = l
+}
 
-	// Use same check as Prometheus.
-	if expr.Type() != parser.ValueTypeVector && expr.Type() != parser.ValueTypeScalar {
+func triggerFallback(err error) bool {
+	return errors.Is(err, parse.ErrNotSupportedExpr) || errors.Is(err, errNotImplemented)
+}
+
+var errNotImplemented = errors.New("not implemented")
+
+func (e *engine) NewExecutor(q storage.Queryable, _ *promql.QueryOpts, expr parser.Expr, start, end time.Time, step time.Duration) (model.VectorOperator, error) {
+	// if step = 0 it is an instant query.
+	// Use same check as Prometheus for range queries.
+	if step > 0 && expr.Type() != parser.ValueTypeVector && expr.Type() != parser.ValueTypeScalar {
 		return nil, errors.Newf("invalid expression type %q for range query, must be Scalar or instant Vector", parser.DocumentedType(expr.Type()))
 	}
 
+	// TODO(bwplotka): We plan to have single plan that can be optimized in both physical and logical context. Rename.
```

Comment:
Would you agree @saswatamcode @fpetkovski ?

Reply:
Just to clarify, this means that for naive plan -> executor -> execute flow, we would do rule-based optimizer step in naive plan, create tree of operators in executor and then execute that?

Reply:
Yes. We will have `plan` which we can do rule based or cost esimtation optimizations. This will give us optimized plan. This can be then transformed to something that can be exected

Reply:
I think the `plan` terminology is getting a bit confusing, especially after the logical plan PR. 😅 

So afaiu theoretically, `logical plan` refers to the query tree and is basically a high-level arrangement of "logical" operators i.e operators without implementations (we don't do this yet, we only have `Plan` interface, which takes in and runs `Optimizers` that traverse and transform nodes in the query tree). This `logical plan` can be optimized via rules which change the relational expression of the query to make the query more efficient and have the engine do less work. With cost-based optimization, we can create several different `logical plans` and pick the one with the least estimated cost (this is also not implemented yet).

From this `logical plan` we can create a `physical plan` (we currently just pass expression with new `logicalplan` nodes), which forms a tree of executable operators that passes actual lower-level storage details to the executable operators. It can be executed by calling `Next()` and `Series()` on root node. Optimizations here include optimized operators which can take advantage of lower-level storage details. This means it can add and implement new operators for intra-operator parallelism, concurrency, flow control, batched selects, etc like `exchange` operators.

> Yes. We will have plan which we can do rule based or cost esimtation optimizations. This will give us optimized plan. This can be then transformed to something that can be executed

So with single `plan` in future, you mean a combined logical + physical plan, that would create an executable operator tree with `exchange` operators as well right from the start, and perform rule-based/cost-based optimizations on it before executing? 🙂 

I think the range/instant query method refactor great! For the combined or split plan, I'm not sure what would be easier to implement and change. 🤔 

Maybe @fpetkovski would have better opinions on this?

Reply:
I think we might be overcomplicating things. Logical operators (AST) = logical plan. Physical operators  = physical plan.
https://learn.microsoft.com/en-us/sql/relational-databases/showplan-logical-and-physical-operators-reference?view=sql-server-ver16

Reply:
Removed this comment to avoid confusion.

Reply:
Yea, I want to do it literally now (control concurrency, distributions etc) - and I blocked by this.